### DISCUSSION TITLE
Roadmap 2025

### DISCUSSION THREAD

**Discussion by aTrotier** [MagneticResonanceImaging/MRIReco.jl#216] [General]
Hi MR community,

I am starting to work on MRIReco.jl more intensively and I will be happy to work on some enhancements. Maybe others are also volunteer for that ?
I think @tknopp don't have a lot of time for development but maybe you have some goals/vision for the repository ?

Some question that might help to define a roadmap :
- Who is using MRIReco.jl either in subpackage like @cncastillo or for their daily routine research job ? 
- What functions are you using ?
- What features / enhancement would you like to see ?

Related to discussion : #149

**Comment by aTrotier** [comment]
For my part I am using MRIReco for both preclinical (bruker) and clinical reconstruction (siemens). I am using MRIReco for the conversion. For advanced reconstruction (subspace / CS...) I generally work with BART to validate my works and then moved on to MRIReco to see if it also works and stick with it if it is the case.

I am starting to work more and more with pypulseq (siemens) and this year I would like to perform an inline reconstruction with gadgetron -> Gadgetron.jl -> MRIReco.jl -> scanner

## Suggested enhancement :

- Siemens data reader (Related to #73 ) : I am either converting the data with `siemens_to_ismrmrd` or with pymapvbvd (called via CondaPkg.jl / PythonCall.jl). Maybe we can use the parameter_map from siemens_to_ismrmrd in order to create the RawAcquisitionData structure correctly filled.
- Associate different trajectory / subsampling for each dimension, right now It is only the case for the echo dimension.
- Long term rework of the dimension available : 
   - more homogeneous between the different object (sensitivity maps is expected has x,y,z,ch but the image after reconstruction x,y,z|sl,echo,ch,rep
   - handle more dimensions (MRD EncodingCounters : average / slice / contrast / phase / repetition/set /segment +/- user)
- Benchmark against BART more frequently (reconstruction time / memory consumed but also **accuracy** see #114 )

**Reply by aTrotier** [reply]
## Copy of our last roadmap discussion : 
1. Fixing the accuracy / convergence issues for CS https://github.com/MagneticResonanceImaging/MRIReco.jl/issues/114
2. implement and benchmark wavelet transformed for non-square / not 2^n matrix https://github.com/tknopp/SparsityOperators.jl/issues/12
3. Model-based reconstruction (moba in BART, PyQMRI)
4. unified EPG package.

**Reply by aTrotier** [reply]
Does the MR community interested in a Julia alternative of qMRLab ? 

I have started to do some stuff in [QuantitativeMRI.jl](https://github.com/aTrotier/QuantitativeMRI.jl) but it in standby and would really need more discussion to plan ahead of time the features / inter-operatibility with other packages

**Reply by cncastillo** [reply]
I have heard about qMRLab, but I am unfamiliar about what they do. It seems like a good idea!

**Comment by nHackel** [comment]
I am starting a new job next year and I am not entirely sure yet how much time I will have for MRIReco and related packages (NFFT.jl, RegularizedLeastSquares.jl and LinearOperatorCollection.jl), but I do have some features planned that I want to do in 2025 that relate to MRIReco.jl:

### NFFT.jl:
 This year I've added GPU-support to the mentioned packages, however the GPU NFFT implementation is still lacking behind the CPU one. In particular, I think it's the preprocessing where the GPU just works with a sparse matrix. Here, @tknopp and I need to take a look if the implementation is feasible and "worth it". Alternatively, there were some other NFFT implementations that have decent GPU support. We might also just make an operator based on that package available to MRIReco.jl.
 
 ### AbstractImageReconstruction.jl
  In our magnetic particle imaging reconstruction package MPIReco.jl, we ran into the issue that the number of reconstruction algorithm variants could not be easily maintained  by a dict or keyword-argument interface. In particular, if we wanted to add a new preprocessing step or a different algorithm all together, we always have to "manually" put int a new branch in the code like:
  
```julia
function reconstruction(data; algorithm = "newAlgo", kwargs...)
  # chose algorithm
  if algorithm == "simple"
    reconstruction_simple(...)
  else if ...
    ...
  else if reconstruction == "newAlgo"
    reconstruction_my_algo(data; kwargs...)
  end
end
```
This also stops other people from being able to extend our reconstructions with variants in their packages since they can't easily add a new branch to our existing methods.

Likewise since keyword arguments don't take part in dispatch, we weren't able to make effective use of Julias multiple-dispatch to handle specific types of processing steps or reconstruction steps. As an example we did have dispatch on the type of reconstruction input, but now thos break down we have two algorithms that are supposed to work for the same type of data.  Another issue is that a flat dictionary can't handle two variables with the same name relating to different processing steps. We had a case where the same argument would need to be applied to the measurement data and the data necessary to build the operator but with different values. All those cases are things we couldn't depict in our existing interface and which would also affect MRIReco.jl
  
This was our motivation behind AbstractImageRecosntruction.jl, which we've also been working on this year. The idea here is that a reconstruction algorithm is a stateful object/struct that implements a reconstruction as a series of `process`-ing steps with a given structure of (nested) parameters. The idea is to have both an algorithm, parameters and the data to dispatch on. New variants of processing steps or even algorithms can then be implemented by adding new structs. This extension of course works across package boundaries.

I've recently added some documentation to AbstractImageReconstruction.jl where I implement a small CT reconstruction framework to show how the interface is intended to be used. We've also changed our MPIReco.jl framework to completely use it and it really helped to quickly prototype and implement new reconstruction algorithms or rather variants of existing algorithms which differ only in very specific processing steps.

The abstract interface also offers a lot of utility functions. For example, one can provide a user a "template" of an algorithm, which is essentially the structure of an algorithm and its parameters where the parameters only have to be partially filled in.. This can be used to provide standard recos such as the ones we have in MRIReco.jl, but also to store the concrete instance that reconstructed an image. These templates also meant to be easily "discoverable" programmatically, so that they can be displayed in the REPL or a GUI Other functionality is caching, where we can transparently cache processing results between reconstructions.

My more concrete plan is to further improve the functionality of the abstract packge and MPIReco.jl. In particular, I want to have a generic reconstruction GUI with tomographic data viewers that dynamically adapts its parameter input to a given plan and I also want to look into support potentially distributed computations. During these implementations the interface should also stabilize a bit more, though we are very happy with what we can do with it already.

I am definitely interested in also extended this interface to MRIReco.jl, but I'm not sure yet if I can provide a proposal this year and it would be a fairly different interface. We wouldn't lose any functionality and gain more options and we could even try to be backwards compatible by mapping the existing calls onto the "new" interface

**Reply by aTrotier** [reply]
AbstractImageReconstruction.jl seems great. It will take me some time to really understand the implication but it kind of looks like the gadgetron / gadgets implementation : https://github.com/gadgetron/gadgetron

Not sure if we should keep the backward compatibility rather than focusing our efforts of keeping the documentation up-to-date :)

**Reply by nHackel** [reply]
A small sidenote to AbstractImageReconstruction.jl. I've recently released the initial version of DaggerImageReconstruction.jl. That package allows us to move whole reconstruction algorithms or just parts of one to a different Julia process. With that I can configure and work with a reconstruction on my laptop and have the actual reconstruction happen on a different machine which features a GPU or just stronger resources in general.

Another setting would be accessing data which is only available in certain machines.

This is all implemented generically, so any reconstruction package implementing AbstractImageReconstruction could get these features for "free"

**Reply by cncastillo** [reply]
It would be really interesting if we can use  AbstractImageReconstruction.jl directly on the scanner with Pulseq and get fully open-source sequences and reconstructions :)!

**Reply by nHackel** [reply]
That would be super cool! I have a bit of a time crunch for a conference atm, but towards the end of march and onwards I could prepare a small presentation about AbstractImageReconstruction.jl and show it to you two @aTrotier, @cncastillo.

More than a year ago I had a briefly talked with Mirco Grosser about migrating MRIReco, but there we mostly talked about just wrapping the high-level `reconstruction_x` methods in their own structs. That is a start, but if we integrate it a bit deeper we can make MRIReco more flexible.

I just don't have the extensive domain knowledge to properly design the integration on my own, but maybe together and potentially some other developers/users we can design a plan. In theory that could also be a GSoC project, but it might be too experimental

**Reply by tknopp** [reply]
I would love to see MRIReco at some point being migrated to `AbstractImageReconstruction`. I personally don't have hacking time but if someone with more MRI knowledge could pair up with @nHackel, this would be really cool.

I think one crucial element would be to define a hierarchy of MR reconstruction algorithms to better understand what type hierarchy one should target there. So one needs to think of algorithm families that share common elements. And that is exactly where input would be needed for Niklas.

**Reply by aTrotier** [reply]
I will be happy to help but I am not sure to have enough knowledge on this topic especially with ML and DL approach. 

Maybe @JeffFessler has some thoughts about that ?

**Reply by JeffFessler** [reply]
Perhaps https://github.com/JuliaNLSolvers/Optim.jl could provide inspiration.  I have not thought about type hierarchies for algorithms.

**Reply by nHackel** [reply]
@cncastillo and @aTrotier, I have not managed to think more about using AbstractImageReconstruction in MRIReco (other than the abstract wish that, that might be cool 😄). But I did have some time to update the MPIReco documentation and that kind of shows how a reconstruction package based on that interface might look like. [Here](https://magneticparticleimaging.github.io/MPIReco.jl/dev/) is the documentation.

Maybe we can schedule a meeting sometime in the upcoming months and I can talk a bit more about it and answer some questions

**Reply by cncastillo** [reply]
Wow very nice documentation! 

> Maybe we can schedule a meeting sometime in the upcoming months and I can talk a bit more about it and answer some questions

That would be great, count me in! :smile: I am currently in California, so I could do 7:00 here any day, that should be 16:00 in Germany.

**Reply by nHackel** [reply]
I've finally had the time to [propose a change](https://github.com/MagneticResonanceImaging/MRIReco.jl/pull/277)

**Comment by aTrotier** [comment]
Should we start again the [MRIReco Community Meeting](https://github.com/orgs/MagneticResonanceImaging/discussions/99) ? Is the community large / active enough ? @mrikasper

I see some activity around KomaMRI @cncastillo :)

**Reply by cncastillo** [reply]
Thumbs up for the meetings!

**Comment by tknopp** [comment]
One question that would be on my list is, if we should put the individual packages into dedicated repositories. In that way one could join the two ISMRMRD implementations (https://github.com/gadgetron/Gadgetron.jl and MRIFiles.jl). Of course that would already possible now but probably it is easier to do once the repositories are split. Ideally everything moves into MRIFiles.jl and Gadetron.jl could then rely on that package.

**Reply by nHackel** [reply]
We can also now do CI-Tests across repositories, so we also don't lose that CI information. Well MRIFiles could see if it breakes something upstream but the upstream wouldn't automatically see that MRIFiles broke, but I think that is good enough

**Comment by cncastillo** [comment]
I think many of these enhancements could be a Google Summer of Code project. I participated this year and it was a very good experience, and a lot of work can be done in the summer (http://juliahealth.org/JuliaHealthBlog/posts/ryan-gsoc/Ryan_GSOC.html). Even if not part of GSoC, many students are looking for programming projects related to medicine. I am sure @thecedarprince would be happy to include some ideas/proposals. Also, if they are math related I think there are some that could be included as a class project for Northeaster University (https://discourse.julialang.org/t/northeastern-university-looking-for-research-projects-for-student-teams/114880/12).

Also, Juliacon 2025 will have a dedicated JuliaHealth mini-symposium to show of the state of Julia for biomedical applications, and I am sure it would be a good place to get some more students interested :) ! It would be very nice to have more people from medical imaging, the abstract submission site will be open soon! I will be there! https://juliacon.org/2025/

Is any of you planning to go to the ISMRM in Hawaii? We could have an informal meeting there, and maybe plan a Julia related member initiated session (https://www.ismrm.org/24m/mis/) for next year (2026). For a while I been meaning to do a MRI-oriented introduction to Julia, I will be doing one in January (https://github.com/LIBREhub/Advanced-Scientific-Computing-Using-Julia/), and maybe something similar can be done at ISMRM, with talks related to the current ecosystem and what is currently possible in Julia, and what needs a few extra hands. 

While is not all only MRIReco enhancements, I would be particularly interested in:
- Koma/MRIReco: Automatic differentiation support (we briefly talked about this with @tknopp  and @nHackel, but I had no time to test it before ISMRM, but I have been experimenting with it)
- Koma: Implementing GPU accelerated EPG or PDG methods
- Pulseq: Finish the Pulseq writer in  https://github.com/JuliaHealth/KomaMRI.jl/pull/284 (Koma)
- Pulseq: Finish the Pulseq helper functions with Unitful.jl support in https://github.com/JuliaHealth/KomaMRI.jl/pull/305 (I have talked to Maxim Zaitsev and Jon-Fredrik Nielsen, to move these with the previous point to a dedicated Pulseq.jl package)
- Pulseq: https://github.com/JuliaHealth/KomaMRI.jl/issues/308

Small issues:
- MRIReco: https://github.com/MagneticResonanceImaging/MRIReco.jl/issues/199
- MRIReco: avoid the need to do `using MRIReco.RegularizedLeastSquares`

**Reply by aTrotier** [reply]
Happy New Year!

Great idea about both the GSoC projects. I'm not entirely sure how much involvement is required from us, but it would be helpful if you could share your past experience during a meeting—specifically, how much involvement is needed, the scope of the projects, and so on.

I won't be attending ISMRM this year (it’s too far from France!), but it’s a great idea to try starting something at ISMRM for the MR Julia community—perhaps something like the BART secret session.

By the way, I need to build an MRI course focused on simulation, Pulseq, and reconstruction. I think most of it will be in Julia, except for the Pulseq part for now. Maybe the examples we create could also be used for an ISMRM initiative.

I’d be happy to help with the following points:
- A Julia-dedicated Pulseq writer, including the MR extension label.
- EPG/PDG simulation capabilities, especially for simulating 3D spoiled sequences. (It seems many issues with komaMRI are related to the increased number of isochromats in the spoiling direction.)

**Reply by aTrotier** [reply]
I don't know how large is the community, but we could also start to do some webinars like [BART](https://www.youtube.com/@bart-toolboxdepartmental7435) or [gadgetron](https://www.youtube.com/playlist?list=PL3dITfhovhLW-ySr_KPBS8KKI9tdLjfdR)

**Reply by cncastillo** [reply]
Sorry I forgot to reply to this. 

>  I'm not entirely sure how much involvement is required from us, but it would be helpful if you could share your past experience during a meeting

I am happy to explain more during a meeting. But basically, I met with my student one hour per week, and talked to him on Slack to help him. 3 hours per week in total maybe? I could dig the GSoC survey where I estimated how much time was spent on my student. But overall it was a **very positive experience**, there are very talented students applying to GSoC and we can later give them recommendation letters if they apply to a job + the GSoC stipend, so overall is a win-win scenario. We should really start planning a few self-contained projects where we need some extra hands and upload them for students to see, as the next GSoC will be soon! 

>  ... it’s a great idea to try starting something at ISMRM for the MR Julia community—perhaps something like the BART secret session.

We should definitely do this. It would be very nice to meet if anyone else is going to Hawaii :). It would be nice to create a logo for the "MR Julia community", we could print stickers and give them during ISMRM. It is a pretty effective way to do some publicity.

> By the way, I need to build an MRI course focused on simulation, Pulseq, and reconstruction. I think most of it will be in Julia, except for the Pulseq part for now. Maybe the examples we create could also be used for an ISMRM initiative.

Can you tell me what would be the most needed functions? It shouldn't be very difficult to get something working for an initial release of a Pulseq.jl package. I have all the pieces I just need to put them together. We could have a GSoC student implement a few and hopefully have Unitful.jl support.

> I’d be happy to help with the following points:
>
>    A Julia-dedicated Pulseq writer, including the MR extension label.
>    EPG/PDG simulation capabilities, especially for simulating 3D spoiled sequences. (It seems many issues with komaMRI are related to the increased number of isochromats in the spoiling direction.)

The Pulseq writer is almost done in https://github.com/JuliaHealth/KomaMRI.jl/pull/284 it just needs an extra push. I have used it already in this work https://archive.ismrm.org/2024/5097.html. 

For the EPG/PDG I really want to include them, but I think I want to focus on finishing the multi-coil implementation first (@Stockless has some preliminary results) and then make the default Bloch simulator differentiable, so we can optimize Pulseq sequences directly.

Maybe we could set a meeting and work together on this!  I am just missing some hands to finish these.

**Reply by aTrotier** [reply]
Thanks @cncastillo 

> We should really start planning a few self-contained projects where we need some extra hands and upload them for students to see, as the next GSoC will be soon!

I will create a dedicated topic.

> It would be nice to create a logo for the "MR Julia community"

Done : https://github.com/MagneticResonanceImaging/MRILogo.jl but it is still open to discussion :)

> Can you tell me what would be the most needed functions? 

My idea for this course is to ask a group of students to build a Quantitative MR sequence (something like a single-slice MP2RAGE for T1 mapping,Variable Flip Angle for T1, MESE for T2, T2 prep sequence) using pulseq. They take a look at the results with Koma and if it works (after I also take a look at the code), we give them access to the scanner to test it.

The missing functions that can help with that are: 
- Labels **++**: that will be used for both kspace position of profiles but also contrast / repetition / slices. They still can do it by hand (like I did) but in the long term even for me it will be easier to not have to worry about that
- EPG : A lot of sequence are gradient echo based and you get wrong results with the default parameters. For experienced user, you know that and you are confident enough that your sequence is suppose to work. But for a student, they will not be able to make the distinction between artefacts from subsampling of the phantom or error in the acquisition / reconstruction pipeline.
- Pulseq.jl : ideally with a CI that validate it with PyPulseq :)

I can work on some of them and we could see if it is possible to propose them for GSoC

**Reply by cncastillo** [reply]
> Done : [MagneticResonanceImaging/MRILogo.jl](https://github.com/MagneticResonanceImaging/MRILogo.jl) but it is still open to discussion :)

Created a new one. What do you think about this one?

<img src="https://github.com/user-attachments/assets/321815a2-582c-4b2b-a065-3d2c79ca915f" width="300"/>

**Reply by aTrotier** [reply]
Nice, I squeezed the icon a little bit.

<img width="1313" alt="Screenshot 2025-02-24 at 09 54 56" src="https://github.com/user-attachments/assets/1d4f9466-ac47-45a3-be9c-0bfe07d63042" />
<img width="1310" alt="Screenshot 2025-02-24 at 09 54 43" src="https://github.com/user-attachments/assets/d659df5f-9292-4989-9ce1-9e0c650bb96e" />

---
<img width="1441" alt="Screenshot 2025-02-24 at 11 56 24" src="https://github.com/user-attachments/assets/31e97a8c-e26a-46bb-8178-576ff931ef3b" />
<img width="1405" alt="Screenshot 2025-02-24 at 09 49 51" src="https://github.com/user-attachments/assets/39eae43a-92a9-4329-a44f-cfa5f0ae0d21" />

Let's vote ?

**Reply by tknopp** [reply]
I like the newest one from @cncastillo  most ( but the other one is also nice! )

**Reply by mrikasper** [reply]
Thank you for your efforts, both of you!
 
I like @cncastillo's version as well, but can we rearrange the colors so that they are in the same way as in the Julia logo triangle (i.e., green on top, red on the left, purple on the right, and maybe the blue as color for the bore)?

**Reply by cncastillo** [reply]
Sure!

<img src="https://github.com/user-attachments/assets/4b03d836-8375-44aa-a97e-42ad63c057e1" width="300"/>

**Reply by mrikasper** [reply]
Beautiful, thank you, @cncastillo !

**Comment by JeffFessler** [comment]
I'll be at ISMRM in Hawaii.  Happy to participate there.  No time to do anything before then sadly...

**Comment by cncastillo** [comment]
Hi everyone! (tagging just a few @nHackel @aTrotier @JeffFessler)

First, I encourage you to submit an abstract to JuliaCon 2026, specifically to the Health Mini Symposium (organized by me and Jacob Zelko):

<img width="1536" height="1024" alt="image" src="https://github.com/user-attachments/assets/1e2c502d-dcd5-4cc8-aa34-5267f5df1d0b" />

---

Second, **Google Summer of Code** is slowly approaching (the proposal deadline is 31st of March). This is a very good opportunity to finish features with the help of very enthusiastic and capable students. If you are interested in participating and you have a 3-month project idea, please let me know! (I might need to figure out the details on how to include proposals coming from MagenticReconanceImaging)

**Reply by aTrotier** [reply]
Hi thanks for the reminder !

I did not work on MRIReco a lot in 2025 but I still use it and might want to include some advanced reconstruction methods like SMS / SMS + parallel imaging but It is really a tricky problem because every sequence (official vs custom) used different convention.

I took a look at what we said last year and something that is really painful to handle is and need to be reworked :

## handle all the dimension of an acquisition
- convertion to raw to acq should be enhanced to store all the standard dimension of MRD (maybe also the USER defined ones)
- All of them should be associated to their own trajectory (right now only the contrast dimension can have different trajectory)

## TWIX reader
Right now I use siemens_to_ismrmd. It works great on my side (I call it in a dedicated conda env) but I am sure having something in Julia will be great !

**Reply by cncastillo** [reply]
Ok, I guess the only thing remaining would be to put the projects in this format:

[GSoC 2024 proposal.pdf](https://github.com/user-attachments/files/24751616/GSoC.2024.proposal.pdf)

The first part is done by the mentor, and the document at the bottom will be submitted by the student.

And after asking about the details to Jakob Zelko (@TheCedarPrince), do a PR like this one: https://github.com/JuliaLang/www.julialang.org/commit/f6155d5b2f7f5d5dbf379118c0ea3d3513f802a8#diff-bffb502c7e528858c17d5903b45d4672dec7a3b6cba9b288167391e5f2284fa2

**Reply by aTrotier** [reply]
Do you think we should focus on only one project ?
Let see if somebody else can propose something. I did not put a lot of thought on this idea.

What do you think is useful to add/fix in MRIReco that can benefits to koma ? @cncastillo

**Reply by nHackel** [reply]
Hmm the only project I'd have in mind would be a proof-of-concept of changing the reconstruction interface to use AbstractImageReconstruction.jl, but I don't think I'll have enough time to properly supervise another student on my own at the moment and I'm not sure if MRIReco would currently benefit from the other interface.

As long as we can still "fit" all our algorithms in the current interface and don't require it to be more extendable/hackable, we are probably fine with leaving things as they are.

I'd also focus on whats interesting for KomaMRI.jl if you have any concrete ideas

**Reply by tknopp** [reply]
@cncastillo: One more general question: Right now we have the sub package [MRISimulation](https://github.com/MagneticResonanceImaging/MRIReco.jl/tree/master/MRISimulation) that we primarily use for generating simulation data that can be used for reconstruction. It is somewhat important that it is a) fast and b) allows to exactly describe extended models like for instance the field-inhomogeneity model, or arbitrary trajectories. My question is: under these constraints: is it possible to do all this in Koma.jl? So, is possible to have "fast paths" for the idealized imaging equations or do you always need a full bloch solver (which I would assume is much more expensive that FFT/NFFTs). Potentially this could be a GSoc.

**Reply by cncastillo** [reply]
Sorry for the delay. I will reply to all comments below:

@aTrotier 
> Do you think we should focus on only one project ?

It doesn't really matter, it can be more than one project. But if you have three projects, each of those need three mentors. So your total dedicated mentoring time = 1 project. 

If you ask me both "handle all the dimension of an acquisition" and a "TWIX reader" seem very useful.

@nHackel 
> I'd also focus on whats interesting for KomaMRI.jl if you have any concrete ideas

Mmm, "handle all the dimension of an acquisition" seem like a good one, as we always have had problems reconstructing 3D data (https://github.com/JuliaHealth/KomaMRI.jl/issues/323) and probably multi-echo, multi-slice, etc.

@tknopp 
> is it possible to do all this in Koma.jl? So, is possible to have "fast paths" for the idealized imaging equations or do you always need a full bloch solver

We could create new simulation methods: "FastGRE", "FastSE", "FastbSSFP", etc. that use the steady-state formulas for each sequence, and gave k-space data using a NFFT. 

The requirements would be:
- field-inhomogeneity model, 
- arbitrary trajectories,
- coil sensitivities,
- am i missing something else?

The only blocker is that I haven't found the time to finish https://github.com/JuliaHealth/KomaMRI.jl/pull/548 (which implements coils). So I could propose a project to add coil sensitivities and "fast paths" for some common sequences.

**Reply by aTrotier** [reply]
Ok, 
are you all agree if we push forward the project :  increase of the dimension handle by MRIReco ? @nHackel @tknopp 
I remember having this discussion and I think one of you wanted to do more abstract in order to handle any new dimension (and not rely on the MRD format

Do you want also to add a summer project in order to finish the multi-coil PR of KomaMRI ?
